File: backend/app/api/emi.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from uuid import UUID
from datetime import date

from app.core import get_db, get_current_user, get_current_admin_user
from app.models import User, EMIContract, EMIPayment, ContractStatus, PaymentStatus
from app.schemas import (
    EMIContractCreate, EMIContractUpdate, EMIContractResponse, EMIContractListResponse,
    EMIPaymentCreate, EMIPaymentResponse, EMIPaymentListResponse, OverdueReport
)
from app.services import EMIService, DeviceService
from app.models import DeviceStatus

router = APIRouter(prefix="/emi", tags=["EMI Management"])


@router.post("/contracts", response_model=EMIContractResponse, status_code=status.HTTP_201_CREATED)
async def create_contract(
    contract_data: EMIContractCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new EMI contract"""
    contract = EMIService.create_contract(db, contract_data, current_user.id)
    return contract


@router.get("/contracts", response_model=EMIContractListResponse)
async def list_contracts(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    status: Optional[ContractStatus] = None,
    customer_id: Optional[UUID] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all EMI contracts"""
    query = db.query(EMIContract)
    
    if status:
        query = query.filter(EMIContract.status == status)
    if customer_id:
        query = query.filter(EMIContract.customer_id == customer_id)
    
    total = query.count()
    contracts = query.order_by(EMIContract.created_at.desc()).offset(skip).limit(limit).all()
    
    return EMIContractListResponse(total=total, contracts=contracts)


@router.get("/contracts/{contract_id}", response_model=EMIContractResponse)
async def get_contract(
    contract_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get contract by ID"""
    contract = db.query(EMIContract).filter(EMIContract.id == contract_id).first()
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    return contract


@router.put("/contracts/{contract_id}", response_model=EMIContractResponse)
async def update_contract(
    contract_id: UUID,
    contract_data: EMIContractUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Update contract details"""
    contract = db.query(EMIContract).filter(EMIContract.id == contract_id).first()
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    update_data = contract_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(contract, field, value)
    
    db.commit()
    db.refresh(contract)
    return contract


# Payment endpoints
@router.get("/contracts/{contract_id}/payments", response_model=EMIPaymentListResponse)
async def list_contract_payments(
    contract_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all payments for a contract"""
    payments = db.query(EMIPayment).filter(
        EMIPayment.contract_id == contract_id
    ).order_by(EMIPayment.installment_number).all()
    
    return EMIPaymentListResponse(total=len(payments), payments=payments)


@router.post("/payments/{payment_id}/record", response_model=EMIPaymentResponse)
async def record_payment(
    payment_id: UUID,
    payment_data: EMIPaymentCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Record a payment"""
    payment = EMIService.record_payment(db, payment_id, payment_data, current_user.id)
    if not payment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Payment not found"
        )
    
    # If payment is made, check if device should be unlocked
    if payment.status == PaymentStatus.PAID:
        contract = payment.contract
        # Check if all overdue payments are cleared
        overdue = db.query(EMIPayment).filter(
            EMIPayment.contract_id == contract.id,
            EMIPayment.due_date < date.today(),
            EMIPayment.status.in_([PaymentStatus.PENDING, PaymentStatus.PARTIAL])
        ).count()
        
        if overdue == 0:
            # Unlock device
            DeviceService.update_device_status(
                db, contract.device_id, DeviceStatus.ACTIVE,
                current_user.id, "Payment received - all dues cleared"
            )
    
    return payment


@router.get("/overdue", response_model=List[OverdueReport])
async def get_overdue_payments(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get all overdue payments with customer and device info"""
    payments = EMIService.get_overdue_payments(db)
    
    reports = []
    for payment in payments:
        contract = payment.contract
        customer = contract.customer
        device = db.query(__import__('app.models', fromlist=['Device']).Device).filter(
            __import__('app.models', fromlist=['Device']).Device.id == contract.device_id
        ).first()
        
        days_overdue = (date.today() - payment.due_date).days
        
        reports.append(OverdueReport(
            contract_id=contract.id,
            contract_number=contract.contract_number,
            customer_name=customer.full_name,
            customer_phone=customer.phone,
            device_imei=device.imei if device else "N/A",
            due_date=payment.due_date,
            days_overdue=days_overdue,
            amount_due=payment.due_amount - payment.paid_amount,
            device_status=device.status.value if device else "unknown"
        ))
    
    return reports


@router.get("/due-soon", response_model=List[EMIPaymentResponse])
async def get_payments_due_soon(
    days: int = Query(3, ge=1, le=30),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get payments due within specified days"""
    payments = EMIService.get_payments_due_soon(db, days)
    return payments


@router.delete("/contracts/{contract_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_contract(
    contract_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Permanently delete a contract and all its payments"""
    contract = db.query(EMIContract).filter(EMIContract.id == contract_id).first()
    if not contract:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Contract not found"
        )
    
    # Delete all payments for this contract
    db.query(EMIPayment).filter(EMIPayment.contract_id == contract_id).delete()
    db.delete(contract)
    db.commit()
    return None

File: backend-node/src/routes/emi.js
import { Op } from 'sequelize';
import { EMIService, } from '../services/emiService.js';
import { DeviceService } from '../services/deviceService.js';
import {
  EMIContract, EMIPayment, Customer, Device,
} from '../models/index.js';
import { serializeContract, serializePayment } from '../utils/serializers.js';
import { isoDate } from '../utils/time.js';
import { httpError, requireUser, requireAdmin } from '../middleware/auth.js';
import { toEnumName } from '../utils/enums.js';
import { ah } from './auth.js';

const CONTRACT_UPDATABLE = ['grace_period_days', 'notes', 'status'];

function todayStr() {
  return new Date().toISOString().slice(0, 10);
}

function daysBetween(fromDate, toDate) {
  const a = new Date(`${fromDate}T00:00:00Z`).getTime();
  const b = new Date(`${toDate}T00:00:00Z`).getTime();
  return Math.round((b - a) / 86400000);
}

export default function registerEmiRoutes(router) {
  router.post('/emi/contracts', requireUser, ah(async (req, res) => {
    const contract = await EMIService.createContract(req.body || {}, req.user.id);
    res.status(201).json(serializeContract(contract));
  }));

  router.get('/emi/contracts', requireUser, ah(async (req, res) => {
    const skip = Math.max(0, parseInt(req.query.skip, 10) || 0);
    let limit = parseInt(req.query.limit, 10);
    limit = Number.isNaN(limit) ? 50 : Math.min(100, Math.max(1, limit));
    const where = {};
    if (req.query.status) where.status = toEnumName(req.query.status);
    if (req.query.customer_id) where.customer_id = req.query.customer_id;
    const { count, rows } = await EMIContract.findAndCountAll({
      where, order: [['created_at', 'DESC']], offset: skip, limit,
    });
    res.json({ total: count, contracts: rows.map(serializeContract) });
  }));

  router.get('/emi/contracts/:contractId', requireUser, ah(async (req, res) => {
    const contract = await EMIContract.findByPk(req.params.contractId);
    if (!contract) throw httpError(404, 'Contract not found');
    res.json(serializeContract(contract));
  }));

  router.put('/emi/contracts/:contractId', requireAdmin, ah(async (req, res) => {
    const contract = await EMIContract.findByPk(req.params.contractId);
    if (!contract) throw httpError(404, 'Contract not found');
    const data = req.body || {};
    for (const field of CONTRACT_UPDATABLE) {
      if (Object.prototype.hasOwnProperty.call(data, field)) {
        contract[field] = field === 'status' ? String(data[field]).toLowerCase() : data[field];
      }
    }
    await contract.save();
    res.json(serializeContract(contract));
  }));

  router.get('/emi/contracts/:contractId/payments', requireUser, ah(async (req, res) => {
    const payments = await EMIPayment.findAll({
      where: { contract_id: req.params.contractId },
      order: [['installment_number', 'ASC']],
    });
    res.json({ total: payments.length, payments: payments.map(serializePayment) });
  }));

  router.post('/emi/payments/:paymentId/record', requireUser, ah(async (req, res) => {
    const payment = await EMIService.recordPayment(req.params.paymentId, req.body || {}, req.user.id);
    if (!payment) throw httpError(404, 'Payment not found');

    if (payment.status === 'paid') {
      const contract = await EMIContract.findByPk(payment.contract_id);
      const overdue = await EMIPayment.count({
        where: {
          contract_id: contract.id,
          due_date: { [Op.lt]: todayStr() },
          status: { [Op.in]: [toEnumName('pending'), toEnumName('partial')] },
        },
      });
      if (overdue === 0) {
        await DeviceService.updateDeviceStatus(
          contract.device_id, 'active', req.user.id, 'Payment received - all dues cleared',
        );
      }
    }
    res.json(serializePayment(payment));
  }));

  router.get('/emi/overdue', requireUser, ah(async (req, res) => {
    const payments = await EMIService.getOverduePayments();
    const reports = [];
    for (const payment of payments) {
      const contract = await EMIContract.findByPk(payment.contract_id);
      if (!contract) continue;
      const customer = await Customer.findByPk(contract.customer_id);
      const device = await Device.findByPk(contract.device_id);
      reports.push({
        contract_id: contract.id,
        contract_number: contract.contract_number,
        customer_name: customer ? customer.full_name : '',
        customer_phone: customer ? customer.phone : '',
        device_imei: device && device.imei ? device.imei : 'N/A',
        due_date: isoDate(payment.due_date),
        days_overdue: daysBetween(isoDate(payment.due_date), todayStr()),
        amount_due: Number(payment.due_amount) - Number(payment.paid_amount),
        device_status: device ? device.status : 'unknown',
      });
    }
    res.json(reports);
  }));

  router.get('/emi/due-soon', requireUser, ah(async (req, res) => {
    let days = parseInt(req.query.days, 10);
    days = Number.isNaN(days) ? 3 : Math.min(30, Math.max(1, days));
    const payments = await EMIService.getPaymentsDueSoon(days);
    res.json(payments.map(serializePayment));
  }));

  router.delete('/emi/contracts/:contractId', requireAdmin, ah(async (req, res) => {
    const contract = await EMIContract.findByPk(req.params.contractId);
    if (!contract) throw httpError(404, 'Contract not found');
    await EMIPayment.destroy({ where: { contract_id: req.params.contractId } });
    await contract.destroy();
    res.status(204).end();
  }));

  // Aggregated stats for the dashboard landing page.
  router.get('/emi/dashboard/stats', requireUser, ah(async (req, res) => {
    const today = todayStr();
    const monthStart = `${today.slice(0, 7)}-01`;

    const [
      totalDevices, lockedDevices, onlineDevices,
      activeContracts, totalCustomers, overduePayments,
    ] = await Promise.all([
      Device.count(),
      Device.count({ where: { status: toEnumName('locked') } }),
      Device.count({ where: { is_online: true } }),
      EMIContract.count({ where: { status: toEnumName('active') } }),
      Customer.count(),
      EMIPayment.count({
        where: {
          due_date: { [Op.lt]: today },
          status: { [Op.in]: [toEnumName('pending'), toEnumName('partial')] },
        },
      }),
    ]);

    const paidThisMonth = await EMIPayment.findAll({
      where: {
        paid_date: { [Op.gte]: new Date(`${monthStart}T00:00:00Z`) },
        status: toEnumName('paid'),
      },
      attributes: ['paid_amount'],
      raw: true,
    });
    const monthlyCollection = paidThisMonth.reduce(
      (sum, p) => sum + Number(p.paid_amount || 0), 0,
    );

    res.json({
      total_devices: totalDevices,
      locked_devices: lockedDevices,
      online_devices: onlineDevices,
      active_contracts: activeContracts,
      total_customers: totalCustomers,
      overdue_payments: overduePayments,
      monthly_collection: monthlyCollection,
    });
  }));
}
